luzer/luzer_args.c

```diff
@@ -0,0 +1,284 @@
+        if (argvp == NULL) {
+                return -1;
+        }
+        l_args->argv = argvp;
+        l_args->argv[l_args->argc] = arg;
+        return 0;
+}
+
+NO_SANITIZE static int
+luaL_get_args_from_cli(lua_State *L, luzer_args *cli_args) {
+	lua_getglobal(L, "arg");
+
+	cli_args->argv = malloc(1 * sizeof(char*));
+	if (!cli_args->argv) {
+                return -1;
+        }
+
+	lua_pushnil(L);
+
+        bool use_cli_args = !getenv(ENV_NOT_USE_CLI_ARGS);
+	/* Zero arg is reserved for program name. */
+        cli_args->argc = 1;
+	while (lua_next(L, -2) != 0) {
+		const char *value = lua_tostring(L, -1);
+		const int key = lua_tointeger(L, -2);
+		lua_pop(L, 1);
+
+                if (key < 0) {
+                        continue;
+                }
+
+                const char *arg = strdup(value);
+                if (!arg) {
+                        return -1;
+                }
+
+                if (key == 0) {
+                        cli_args->argv[0] = (char*)arg;
+                        continue;
+                }
+
+                if (use_cli_args) {
+                        if (add_arg_with_mem_allocation(cli_args, (char*)arg)) {
+                                return -1;
+                        }
+                        cli_args->argc++;
+                }
+	}
+	lua_pop(L, 1);
+	return 0;
+}
+
+NO_SANITIZE static int
+luaL_get_args_from_table(lua_State *L, luzer_args *table_args) {
+        if (lua_istable(L, -1) == 0) {
+                return -2;
+        }
+
+	lua_pushnil(L);
+
+	/* Processing a table with options. */
+	table_args->argc = 0;
+        while (lua_next(L, -2) != 0) {
+		const char *key = lua_tostring(L, -2);
+		const char *value = lua_tostring(L, -1);
+		lua_pop(L, 1);
+
+                /* +1 for null terminated */
+                size_t arg_len = FLAG_FULL_SIZE(key) + strlen(value) + 1;
+		char *arg = calloc(arg_len, sizeof(char));
+		if (!arg) {
+                        return -1;
+                }
+		snprintf(arg, arg_len, FLAG_PATTERN_KEY_VALUE, key, value);
+
+		if (table_args->argc > 0) {
+                        if (add_arg_with_mem_allocation(table_args, arg)) {
+                                return -1;
+                        }
+		} else {
+			table_args->argv = malloc(1 * sizeof(char*));
+			if (!table_args->argv) {
+                                return -1;
+                        }
+                        table_args->argv[table_args->argc] = arg;
+		}
+		table_args->argc++;
+	}
+	lua_pop(L, 1);
+	return 0;
+}
+
+NO_SANITIZE static int
+merge_args(luzer_args *cli_args, luzer_args *table_args, luzer_args *total_args) {
+	/* Program name is first argument. */
+        total_args->argc = 1;
+	total_args->argv = malloc(sizeof(char*));
+	if (!cli_args->argv) {
+                return -1;
+        }
+
+	/* Program name on zero index. */
+	total_args->argv[0] = cli_args->argv[0];
+
+	char *corpus_path = NULL;
+	for (int i = 0; i < table_args->argc; i++) {
```

Comment:
Probably it is better to put `table_args->argv[i]` to a separate variable (for example `cur_arg`) at the beginning of loop iteration. This change will reduce length of lines a bit. Currently, it is difficult to read lines in a loop body.